feat: Add version 2 of tables

bendyorke opened the pull request:
Opening this PR for discussion purposes. The code is very much in a WIP stuff and I will greatly refactor it once we settle on the design (probably every line of code will get rewritten into a testable manner)

This PR adds new table support for markdown tables, query tables, and more. It's also introduces a new namespace for the component library, currently named `shui`.

## First of all, some screenshots

<img width="1840" alt="Screenshot 2023-02-01 at 13 58 05" src="https://user-images.githubusercontent.com/3488879/217069996-6acc5c1c-72e2-480a-a174-32c371bf12f9.png">

<img width="1139" alt="Screenshot 2023-02-06 at 13 44 09" src="https://user-images.githubusercontent.com/3488879/217070021-fa5a472a-4395-412e-b2a8-899382d02971.png">


## There are several technical things that this pull request introduces:

1. **Implementation of a component system**: In order to make sure the table is future proof, I've started work on a component library. This work starts with the `table` and `inline` component (although the latter is just a wrapper for now).
2. **Implementation of custom properties via namespaces**: Ideally, we settle on a default theme that looks good for tables, and add a few UI elements to eventually do the customization. For now, the configuration is handled on the block level. @tiensonqin suggested adding default configuration to the `config.edn`, although I still need to do that
3. **Implementation of component versioning**: In efforts to help provide a bug free upgrade path, I've baked versioning into this feature. More work needs to be done to bake this into `shui` components by default, but first I wanted to open it up for discussion.
4. **Implementation of radix color theme**: Currently I'm whitelisting all of the radix color theme variables which is not ideal. This PR does allow customization of colors using radix's color theme. I will be working with Jakob this month (during the rollout of this feature) to get radix colors themes available on Logseq. For now, it allows for the coloring of tables on a table by table basis.

## Main questions I have about this feature:

1. Does the name shui bother anyone? It is Chinese for water, and oh so very cleverly has the name `UI` in it. I don't know if this is a very strange name for the Chinese team, I thought it was a clever cross-cultural pun but I want to give anyone a chance to veto it.

2. How do people feel about the current way to version table components (adding custom props to the block itself)?  How should we handle reserve keywords? 

3. How should we handle testing of the component library? Currently, these are all displayed on lambda, but I can write integration tests for them. I could also set up snapshot testing for the UI library, or try to abstract as much logic as possible to pure functions and unit test them. Does anyone have any strong preferences either way?

## Plans for rollout

1. We will be deciding on a final design direction this week
2. I will clean up and test the code until it is sufficient to merge into master
3. Users will not see any changes unless they add `table/version:: 2` to a block property (or to global config). This gives our nightly users, testing audience, and anyone we tell the chance to test it
5. Fix any bugs that come in for the rest of the month
6. Switch the default table/version to 2 - any users that have pressing issues can set `table/version:: 1` for a time being
7. Eventually remove version 1 (we should discuss how long to keep version 1 around for)

Bad3r commented:
This is awesome!

- how can I swap between table 'types'? i.e. how can I specify a 'scrollable' table or an Incongruent table.
- I am having an issue with horizontal scrolling; seems like the scrollbar gets out of boundary. Also, it's not possible to click on it.

https://cdn.hacklog.sh/vid_2023_02-06_18_44_36.mp4

https://user-images.githubusercontent.com/25513724/217112590-66012edf-e3e1-4f4d-bca4-1954bf42715d.mp4


table used for testing:
```md
#### Usage Summary
|FIELD1|hex    |Dark ambiance                                  |Bright ambiance                               |FIELD5|R  |G  |B  |
|--|--|--|--|--|--|--|--|
|0     |`#2e3440`|background                                     |plain text, caret, brackets                   |N     |46 |52 |64 |
|1     |`#3b4252`|gutters, popups                                |                                              |i     |59 |66 |82 |
|2     |`#434c5e`|current line, selection                        |                                              |t     |67 |76 |94 |
|3     |`#4c566a`|line numbers                                   |                                              |e     |76 |86 |106|
|4     |`#d8dee9`|caret, variables, constants, attributes, fields|gutter, popups                                |S     |216|222|233|
|5     |`#e5e9f0`|                                               |current line, selection                       |n     |229|233|240|
|6     |`#eceff4`|plain text, brackets                           |background                                    |w     |236|239|244|
|7     |`#8fbcbb`|classes, types, primitives                     |classes, types, primitives                    |F     |143|188|187|
|8     |`#88c0d0`|declarations, calls                            |declarations, calls                           |r     |136|192|208|
|9     |`#81a1c1`|syntax, keywords, operators, tags, punctuation |syntax, keywords, operators, tags, punctuation|s     |129|161|193|
|10    |`#5e81ac`|pragmas, comment keywords, pre-processor       |pragmas, comment keywords, pre-processor      |t     |94 |129|172|
|11    |`#bf616a`|errors                                         |errors                                        |A     |191|97 |106|
|12    |`#d08770`|danger, annotations, décor                     |danger, annotations, décor                    |u     |208|135|112|
|13    |`#ebcb8b`|warn, escape, regex                            |warn, escape, regex                           |r     |235|203|139|
|14    |`#a3be8c`|strings                                        |strings                                       |o     |163|190|140|
|15    |`#b48ead`|numbers                                        |numbers                                       |r     |180|142|173|
|16 |`#616e88`|comments                                       |comments                                      |a     |97 |110|136|
table/version:: 2
```

bendyorke commented:
Thanks for testing @Bar3r! This branch wasn't really meant to be tested (it's an old branch) so I hope the experience wasn't too bad for you. This is mainly just a place to host the discussion about the points I mentioned :)

1. There aren't actually different types of table outputs, just different types of table inputs. The table itself will render columns different sizes if there are large discrepancies in the widths of the columns, and will render itself scrollable if there is too much content to fit in a single block
2. This branch is a little bit out of date - that will be fixed in the final version :)

Bad3r commented:
@bendyorke thank you for clarifying! 
minor note; I personally always prefer my tables to have internal borders;
could that be a configurable option/prop?

bendyorke commented:
@Bad3r yes, there is already an option for `table/borders:: true` :) Here are all the current configuration options:

<img width="1402" alt="Screenshot 2023-02-07 at 15 31 36" src="https://user-images.githubusercontent.com/3488879/217273298-bcb0795b-5adf-4ebb-96dd-6de0e2220c6a.png">

All the documentation will be added, currently it's on an internal doc but it will go into the public doc when it gets released

cannibalox commented:
about `Implementation of custom properties via namespaces` : should these custom properties have a distinctive prefix/symbol ? eg `.table/`,`-table/` or `$table/` rather than just `table/` ? 
it would be easier to distinguish them from standard props and avoid collision with existing props, also making it easier to globally hide or list all custom props ?  

also is there any way to have columns resizers ?

bendyorke commented:
@cannibalox that's one of the discussions we are having internally. I'm not sure whether people namespace their props already - so it's possible that just putting them behind a namespace is making them distinct. 

I'd like to avoid too much special symbols, as I don't want them to look too intimidating to the average user. 

At the moment we have a lot of special props that aren't really signified in anyway. I think we started using a `logseq` prefix for some of them, but that isn't entirely enforced.

Another option would be to just use a single namespace for all the styling - something like `view/table-cols::` instead of `table/cols::`. Or `view.table/cols::` would also be an option, but I don't want users to have to remember between dots or hyphens, and I think `view/table-cols::` is cleaner than `view-table/cols::`.

Apologies this was a little bit rambly, but I'd be curious to hear your thoughts on any of those directions?

Bad3r commented:
@bendyorke that's correct; only few are prefixed w/ `logseq.`

here is the list of properties used by Logseq https://github.com/logseq/docs/issues/79

bendyorke commented:
I’ve moved the properties to the `view/` namespace, as well as cleaned up the code. It should be more readable now.

I also added some default gradient helpers in tailwind. These are causing the build time of css to slow to a hault, so I need to look into that.

Refactor isn’t finished, there’s still a couple items that haven’t been reimplemented, but they shouldn’t change the general structure of the code too much.

Biggest things left are testing and theming, but I will create a more detailed list this evening of things left to do.

logseq-cldwalker commented:
@bendyorke This is looking great! Sorry for the delay. Some feedback on above comments:

> For now, the configuration is handled on the block level. @tiensonqin suggested adding default configuration to the config.edn, although I still need to do that

I'd second that suggestion. Personally I only want to set a version once in the config. If we're giving users the option to configure it per component, the behavior starts to feel more like a named version (style) than a numeric version. For examples of configs that only allow a couple values see `:preferred-workflow` or `:file/name-format`

> Does the name shui bother anyone? It is Chinese for water, and oh so very cleverly has the name UI in it. I don't know if this is a very strange name for the Chinese team, I thought it was a clever cross-cultural pun but I want to give anyone a chance to veto it.

I do like me a good pun but will defer this to Chinese teammates

> How do people feel about the current way to version table components (adding custom props to the block itself)?

I would hope we would build all table options including versions as ui components rather than raw text. This is how the query table behaves and is generally the direction we want to be going. Text interfaces to features we have e.g. config.edn and queries have gotten feedback as giving a hacker interface and being less user friendly. I do realize this ask is more expensive so maybe that's something for a later iteration. It would be nice long term if we only support a text or UI interface.

> How should we handle reserve keywords?

Adding built-in properties is thankfully easy nowadays as it's configurable from https://github.com/logseq/logseq/blob/master/deps/graph-parser/src/logseq/graph_parser/property.cljs. It'd be good to prefix all the new keywords with `:logseq` e.g. `:logseq.table/version` to keep with conventions we started back in #6105. Logseq properties live in user space and so it benefits the user and the ecosystem as whole if the platform and plugins stay in their namespaces. If we can't set an example of namespacing under `:logseq`, then we should expect our community to mimic us and cause clashes randomly. If typing is a concern, autocomplete makes this easy e.g. `::ver[TAB]` to type out `:logseq.table/version::`

> How should we handle testing of the component library? Currently, these are all displayed on lambda, but I can write integration tests for them. I could also set up snapshot testing for the UI library, or try to abstract as much logic as possible to pure functions and unit test them. Does anyone have any strong preferences either way?

Since you, @sprocketc and @xyhp915 are the creators of the UI library I think your opinions matter the most. I'm happy to be a tie-breaking vote or facilitate discussion as needed

> :view/table-cols

Version 2 query tables will need to be able to fall back to `:query-properties` as everyone's data is stored that way

xyhp915 commented:
> How should we handle testing of the component library? Currently, these are all displayed on lambda, but I can write integration tests for them. I could also set up snapshot testing for the UI library, or try to abstract as much logic as possible to pure functions and unit test them. Does anyone have any strong preferences either way?

  > Since you, @sprocketc and @xyhp915 are the creators of the UI library I think your opinions matter the most. I'm happy to be a tie-breaking vote or facilitate discussion as needed

I wonder if it is possible to move the UI components to a separate UI Library, and then the logseq repo directly references the packaged component library. Regarding the basic configuration related to the Logseq design system (tailwind.config.js), is it possible to encapsulate it into the logseq UI Library and then import it to the logseq repo for use as a plugin?

sprocketc commented:
> Does the name shui bother anyone? It is Chinese for water, and oh so very cleverly has the name UI in it. I don't know if this is a very strange name for the Chinese team, I thought it was a clever cross-cultural pun but I want to give anyone a chance to veto it.

I like it, but I don't think you should count on my opinion :smile: 

> How do people feel about the current way to version table components (adding custom props to the block itself)? 

I agree with @logseq-cldwalker on this. I don't think it should be a block level option.

> How should we handle testing of the component library? Currently, these are all displayed on lambda, but I can write integration tests for them. I could also set up snapshot testing for the UI library, or try to abstract as much logic as possible to pure functions and unit test them. Does anyone have any strong preferences either way?

Any kind of testing would be appreciated. I have a personal preference for integration tests, because they promote testing the actual behavior, and are a better fit for validating the functionality of UI components. Unit testing logic when possible is nice. Snapshot testing would be cool, but I think a solution like Storybook might be a better option, since it would also serve as the component documentation when we start exporting those components.

bendyorke commented:
Forgot to post the final checklist, as it stands from today:

- [x] `:view/table-cols`
- [x] `:view/table-headers`
- [x] `:view/table-hover`
- [x] `:view/table-compact`
- [ ] `:view/table-version` 
- [x] add group separation
- [x] fix border radii
- [ ] read view config from `config.edn`
- [ ] read query config from `:query-properties`
- [ ] fix mouse interactions (click)
- [ ] fix scrollbar + padding issues
- [ ] integrate with current theming / fix grays on backgrounds 
- [ ] remove flash on initial render

Let me know if I'm missing anything, this should be good to go pretty soon

logseq-cldwalker commented:
> 100% the plan is to do this. However, we need a place to store that data, and block level properties seems as good a place as any. So plan is to support the values here, then config.edn, then work to get a proper UI added to manage them so a user never sees them.

That's great to hear 🙌 

> @logseq-cldwalker I moved everything into a view/table-* prefix. Given that I foresee a lot of view based properties configurable here, I don't want it to appear too obtuse to the user. If it's acceptable to reserve the view namespace (without the logseq. prefix) then I would prefer this solution :)

How a property looks doesn't have to come at the expense of how we store data. We [already have a way to hide certain properties on render](https://github.com/logseq/logseq/blob/4d55fc4b13043fc877fa96967f3f40936e4eed4a/src/main/frontend/components/block.cljs#L2065) or you could just display them by their keyword name e.g. `table-cols`. The users that may find `logseq...` obtuse are also the ones that want the proper UI so this won't effect them in the long run :)

logseq-cldwalker commented:
> I wonder if it is possible to move the UI components to a separate UI Library, and then the logseq repo directly references the packaged component library. Regarding the basic configuration related to the Logseq design system (tailwind.config.js), is it possible to encapsulate it into the logseq UI Library and then import it to the logseq repo for use as a plugin?

@xyhp915 brings up an interesting point about working with https://github.com/logseq/ui. I believe our long term goal with the component system is to make it usable within logseq and within the plugin ecosystem as well. This likely means that components will need to be written in JS. Obviously not something to do in this PR but something to be discussing and coordinating

> I would advise strongly against this - it creates a massive amount of overhead and creates a lot of issues when it comes to versioning, linking, deployments, etc.

I also agree that having this as an external dep would create a lot of friction. Given how much of a mono-repo Logseq already is, it seems reasonable to put in a separate directory. @bendyorke It could be useful to put this new ui namespace in it's own library under `deps/`. It would help to keep the components free of app-specific concerns and would allow you to easily define project-specific workflows e.g. [graph-parser's workflow](https://github.com/logseq/logseq/blob/master/.github/workflows/graph-parser.yml)

cannibalox commented:
sorry if I add noise in the convo.... It's a great read.
@bendyorke thanks for sharing your thoughts. I had prepared a long reply but most of my thoughts were expressed in a better way by @logseq-cldwalker

> @cannibalox that's one of the discussions we are having internally. I'm not sure whether people namespace their props already - so it's possible that just putting them behind a namespace is making them distinct.

once the PR is merged, the concept of namespaced props will probably inspire user's worklflows anyway ? seeing the devs officially supporting the namespaced props will trigger adoption for custom namespaced properties, so I doubt namespacing alone is enough to warrant that sytstem/props are easily distinguishable in the long run.

> I'd like to avoid too much special symbols, as I don't want them to look too intimidating to the average user.
>At the moment we have a lot of special props that aren't really signified in anyway. I think we started using a logseq prefix for some of them, but that isn't entirely enforced.

I agree that symbols as prefix are not intuitive and too abstract (see the tiddlywiki's props below...), however as a user, I find it adds a lot of friction to have 'utility' properties mixed in with my own user-defined properties : 
- user's custom properties are metadata that convey meaninful info related to the users notes
- system props are toggles or settings used by logseq to drive how the data is rendered, organized or processed
so I would love a way to filter out/hide those system-props - even better if there is a dedicated UI opr if those props are moved in config.edn (I'm willing to sacrifice granularity for tidyness in my notes and I guess custom.css would still enable picky users to fine-tune the look of their tables).

So I think all suggestions from @logseq-cldwalker make a lot of sense.
In the long term a dedicated UI should be the way to go.

Ultimately,  I would be great to have a meaningful separation between user-properties and system-properties, either in the UI or/and the naming convention.
use-case : when I share my notes, I probably want to **keep** my user-defined properties (eg: `assigned-to:: someone`, `status:: wip`, but would want to hide/remove utility props (`table/border:: 1px`, `view/background-iumage: imageurl`)...
until a dedicated UI is available, there should be proper way to filter out system properties. 

In this regard, from my lambda user's perspective, the `logseq.` prefix sound like a better solution than just `view/...::` ot `table/...::`

that also brings up the question of the taxonomy/naming convention for logseq props :

| Logseq                 | Tiddlywiki                          |
|------------------------|-------------------------------------|
| collapsed              | $:/tags/AboveStory                  | 
| id                     | $:/tags/AdvancedSearch              |
| created-at             | $:/tags/AdvancedSearch/FilterButton |
| query-properties       | $:/tags/Alert                       |
| query-sort-by          | $:/tags/BelowStory                  |
| query-sort-desc        | $:/tags/ControlPanel                |
| custom-id              | $:/tags/ControlPanel/Advanced       |
| background-color       | $:/tags/ControlPanel/Appearance     |
| ls-type                | $:/tags/ControlPanel/Info           |
| hl-page                | $:/tags/ControlPanel/Saving         |
| hl-stamp               | $:/tags/ControlPanel/Settings       |
| hl-color               | $:/tags/ControlPanel/Toolbars       |
| logseq.macro-name      | $:/tags/EditorTools                 |
| logseq.macro-arguments | $:/tags/EditorToolbar               |
| logseq.tldraw.page     | $:/tags/EditPreview                 |
| logseq.tldraw.shape    | $:/tags/EditTemplate                |
|                        | $:/tags/EditToolbar                 |
|                        | $:/tags/Exporter                    |
|                        | $:/tags/Filter                      |
|                        | $:/tags/Image                       |
|                        | $:/tags/ImportPreview               |
|                        | $:/tags/KeyboardShortcut            |
|                        | $:/tags/Layout                      |
|                        | $:/tags/Macro                       |
|                        | $:/tags/Macro/View                  |

Tiddlywiki's syntax is ugly and **not** intuitive (...maybe $ stands for $y$tem ?) but it also brings benefits with autocompletion: eg typing `$` (or `$:/` if needs to match 3 characters min) would trigger a list of all system tags, ** something which is currently not possible in logseq - you can't easily get a list of all system props.**
TW benefit : you only need to memorize `$` = `system` then rely on autocompletion
TW problem : it's non-intuitive, it's abstract.

Logseq: system properties are a hidden knowledge, the properties names are not following any standard, hard to memorize, hard to find using the autocomplete box.
some have a `ls-` prefix, some `hl-`, others, `logseq.` now maybe `view/` or `table./`... 

it would be great to enforce consistency and future-proofing in the naming convention used for props.
it seems you are going for `logseq.feature-name.property` ? as in 
`logseq.tldraw.page`
`logseq.tldraw.shape`
so that would imply something in the line of :
`logseq.table.border-width`
`logseq.table.background-color`
...

my humble suggestion is to stick with `feature-name` that is used in the documentation, eg : `logseq.tldraw.page` should have been `logseq.whiteboard.page` or `logseq.wb.page` (b/c the lambda user is **not** supposed to know that tldraw = whiteboard). 
even though in the long run, this would be hidden behind a dedicated UI, using the actual feature name is more intuitive than 'tldraw'.
 
> Another option would be to just use a single namespace for all the styling - something like view/table-cols:: instead of table/cols::. Or view.table/cols:: would also be an option, but I don't want users to have to remember between dots or hyphens, and I think view/table-cols:: is cleaner than view-table/cols::.

`view/props::` or `logseq.view/props` imply that namespaces prefixes are based on  "action" like 'view', 'query', 'filter' rather than based on 'feature' (like whiteboard, table, etc).
Either way, I think it should be consistent across all system properties : prefix=action or prefix=feature but don't mix and match.

looking at the exiting system props, the prefixes currently have different 'types' : some are determined by the function name, others by the action, the context...
 eg:
 `ls-` stands for logseq (the context)
 `hl-` is highlight (a feature)
 `logseq.` is another prefix for `ls-`
 `query-` is all related to queries, so function
 `view/` would encompass all display settings across all features, like tables, whiteboards, etc ? `view/table-border::`, `view/whiteboard/bg-color::`, etc...

 as a user, I find the most intuitive and practical taxonomy to be `logseq/feature-name/property` or `.ls/feature-name/property` (just because `.ls/` is less likely to appear in a user's own notes than `logseq/` - which I bet appear in a lot of users' graph).

 sorry for the wall of text... also, I really like the new tables look :)

bendyorke commented:
> @xyhp915 brings up an interesting point about working with https://github.com/logseq/ui. I believe our long term goal with the component system is to use it within logseq and offer it to our plugin ecosystem. This likely means that components will need to be written in JS. Obviously not something to do in this PR but something to be discussing and coordinating

Yeah, our component library needs to handle a couple things from what I've gathered:
1. Support new radix color theming
2. Look and feel more fresh (better native UI)
3. Work on mobile
4. Provide primitives that fulfill all of the first three to plugin authors so they can develop plugins with a consistent feel.

I think these more or less need to be tackled in order. We should build some native components that we use throughout the app that look and feel nice and fit with the future design direction. Then we can abstract out some primitives and provide them to plugin authors.

Likely, this can all be done in cljs - perhaps at the very end, we create a small typescript project that wraps the exports so it's easier to integrate with typescript projects, but I'm not entirely sure that will even be necessary. I know I've exported reagent components and used them in a react project, granted I've never used rum before but I'm not concerned about part of the process. For now, I'd like to get some internal native components that look nice, function well, and can be properly configured (also likely eventually providing plugin authors 'native' like places to put their plugins), and I think the best way to do that is to keep it as close to the code as possible.

@cannibalox thank you for your in depth message! I don't know which part to quote haha. But to both you and @logseq-cldwalker - I've moved all the props now to be `view/table-borders:: true`. This pattern could be used to extend to whiteboards (`view/whiteboards-borders:: true`) or any other component. (note: you cannot have multiple namespaces within a single key).

If this seems too ambiguous for whatever reason, or if view is not the right namespace, we can change it. I can move things over to `logseq/table-borders:: true` but that feels like I'm editing a global logseq property, it doesn't feel specific to the block. Internally, it would just be a mess if I used a `block/table-borders:: true`

These are properties that affect how it is rendered. I'd really like to keep these readable, even if most users don't see them because we eventually have a UI. Some users will interact with them on a text basis, and I think it's reasonable to reserve at least one namespace for all block-level-rendering-properties.

Given that `block` & `render` are already concepts, I went with `view`. For the reasons mentioned already, I'm not a fan of `logseq/table-borders::` but if everyone else likes that better then we can go with that. I'm open to suggestions, let me know your thoughts

bendyorke commented:
Also, we could use the dot notation like you mentioned  (`logseq.table.headers:: uppercase` or `view.table.headers:: uppercase`). I think this is also a clean enough option. I slightly prefer the namespaces, only because it maps a little better with the Clojure under the hood, but the dot notation allows for multiple levels of taxonomy so I'm flexible either way.

tiensonqin commented:
> Likely, this can all be done in cljs - perhaps at the very end, we create a small typescript project that wraps the exports so it's easier to integrate with typescript projects, but I'm not entirely sure that will even be necessary. I know I've exported reagent components and used them in a react project, granted I've never used rum before but I'm not concerned about part of the process. 

@bendyorke That's awesome! Building the UI components in Clojurescript is fine as long as those components can be used in Javascript smoothly. It'll be nice to build a demo (maybe a button component) in Clojurescript and ensure that Javascript can use it. 

> It could be useful to put this new ui namespace in it's own library under deps/.

I agree with this too.

logseq-cldwalker commented:
> Also, we could use the dot notation like you mentioned (logseq.table.headers:: uppercase or view.table.headers:: uppercase). I think this is also a clean enough option. I slightly prefer the namespaces, only because it maps a little better with the Clojure under the hood, but the dot notation allows for multiple levels of taxonomy so I'm flexible either way.

Using dot notation for these new properties e.g. `logseq.table.headers` sounds good if that's what keeps them prefixed with `logseq.`

bendyorke commented:
I've moved the project over to reps and changed the properties to follow the format `logseq.table.version:: 2`. I'm working on tests now.

Just to confirm, this means that in order to change the color of a single block, users will enter:

```
logseq.color:: red
```

To me, this sounds like it would be setting the color of logseq, not a block's color. I do understand the principle behind keeping the `logseq.` prefix, but to me this is very unclear and confusing -- both from a users perspective and the programmers perspective. If you think it is a sufficient tradeoff to remain consistent then I will keep it. 

Please let me know @logseq-cldwalker

logseq-cldwalker commented:
> To me, this sounds like it would be setting the color of logseq, not a block's color. I do understand the principle behind keeping the `logseq.` prefix, but to me this is very unclear and confusing -- both from a users perspective and the programmers perspective.

@bendyorke If this is confusing for the programmer then please reconsider `logseq.X.color` where X is whatever makes most sense to you. As for the user, I've already pointed that tab completion will handle most of the typing and that the render view is customizable. We've also heard from a power user above who preferred the `logseq.` prefix.

> If you think it is a sufficient tradeoff to remain consistent then I will keep it.

As long as it's under `logseq.`, feel free to add whatever you think makes sense

bendyorke commented:
@logseq-cldwalker @Bad3r if we intrafix `logseq.color` with another modifier, then that same modifier should be applied to all other view related properties as well - so if we have `logseq.view.color` then we should have `logseq.view.table.version` which I think is getting unnecessarily long.

I understand the desire for a common namespace, however I can't imagine that the sticking point was that 'not all logseq properties use a single namespace', as much as it was 'logseq properties are not namespaces in a common fashion'. I personally think that if we have `logseq.` properties for internal properties that work under the hood, and `view.` properties that affect how a block is rendered, and we continue to adhere to this, that would appease the power users. If anything, I would assume power users would appreciate the distinction between functionality and display related properties.

However, I don't want this to become a sticking issue for this PR. I feel I'm making enough assumptions in this PR, so I don't want to be stubborn about this issue. IIf you'd like, I'm happy to set up another twitter/discourse/discord poll if you'd rather go that route.

So far the options we have are:

```
logseq.table.version:: 2
logseq.color:: red
```

```
logseq.view.table.version:: 2
logseq.view.color:: red
```

```
view.table.version:: 2
view.color:: red
```

@Bad3r I would try to stay away from using the word block here, only because blocks already means so much in the codebase, and we have several different kinds of block properties already, I don't want to continue to overload that namespace

alxlg commented:
My 2 cents on the properties naming issue (it can be a bit off topic):

In the spirit of cooperation with other projects that could read/write properties (other apps but also export tools and other kinds of integrations), I think it makes sense to adopt something like this:

1. Use the convention that all property keys starting with a dot (`.`) are hidden from the UI (and maybe shown on demand to inspect a block/page?). This is inspired by the well-known Unix standard for hidden files.
2. Property keys that makes sense to be read/write for other project could have no prefix, for example:
    - `.id::`
    - `.collapsed::`
    - `.icon::`
    - `.alias::`
3. Add a prefix like `ls` for Logseq specific properties:
    - `.ls.table.v::` for table version
    - `.ls.heading:: true` for autoheading
4. Ask to plugin devs (and eventually other projects) to do the same to store metadata the user is not supposed to view/edit manually:
    - `.zot.uri::` hypothetical URI stored by a Zotero integration
 5. Make the users free to use whatever property key they want without the dot, including `title::`, `icon::` etc and avoid any conflict with "special" properties.

cannibalox commented:
@bendyorke if I had to pick one option, I'd go with 
```
logseq.view.table.version:: 2
logseq.view.color:: red
```
but ideally, I would favor this taxonomy template `.logseq.%feature%/%property%`, ie:
```
.logseq.table/version:: 2
.logseq.table/color:: red
.logseq.table/border-width:: 2px
.logseq.table/col-width:: 300px 

fantasy properties : 
.logseq.text/color:: blue
.logseq.text/background:: white
.logseq.header/size:: medium 
... (probably some settings would overrule others ?)
```
- this would address @alxlg 's suggestion (I also think starting with `.` is a good way to quickly tell apart system props from user-props)
- this keeps the logseq prefix and should work well with autocomplete (`logseq.` would list all system props, `.%feature%` would filter all related properties to the feature while giving context)
- it should work for any future feature/property combination without overlaps or edge-case 
- even though this produces long names, autocmoplete should alleviate the issue, then once a dedicated UI is in place, this won't matter anyway.

lef-est commented:
> I'm not sure whether people namespace their props already - so it's possible that just putting them behind a namespace is making them distinct.

Have you tested with org-mode? I once changed a property name to a namespaced one and it was very problematic didn't. Bugs everywhere. So I spent a lot of time changing the results back.😂

So please do not use namespace (at least for the first version).

tiensonqin commented on deps/shui/src/logseq/shui/table/v2.cljs:
```diff
@@ -0,0 +1,462 @@
+                          (set-cell-focus [(inc col-index) row-index]))
+           nil)
+         ;; Prevent default actions when the table handles it itself
+         (.preventDefault e))))
```
One minor bug:
Press up/down/left/right will select or open a relative block. Should we disable the global shortcuts here(disable-all-shortcuts)? `stopPropagation` could be helpful too.

bendyorke replied:
I think the keyboard interactions need some love. We should probably disable all shortcuts when a cell is focused, but at the moment that would mean introducing frontend modules into shui, which we're trying to keep abstracted.

I think this is something that should be fixed before making v2 the default, but I think it is too much to fix in this initial PR. If it's a major sticking point for you I can fix it first thing after merge.

tiensonqin commented on deps/shui/src/logseq/shui/table/v2.cljs:
```diff
@@ -0,0 +1,462 @@
+    ; --ls-tertiary-background-color: #f2f2f3;
+    ; --ls-quaternary-background-color: #ebeaea));
+
+(defn lsx 
```
nitpick: `(lsx :text 1)` instead of `(lsx "text" 11)`

tiensonqin commented on e2e-tests/shui/table.spec.js:
```diff
@@ -0,0 +1,304 @@
+import { expect } from '@playwright/test'
```
❤️

tiensonqin commented on src/main/frontend/components/query_table.cljs:
```diff
@@ -165,52 +167,66 @@
+          result-as-text (for [result-entry result'] 
+                           (for [column-name columns] 
+                             (case column-name 
+                               :block (or (get-in result-entry [:block/page :block/original-name])
```
`:block` and `:page` reversed?

tiensonqin replied:
btw there's a bottom border for the table.
<img width="2451" alt="CleanShot 2023-04-03 at 22 38 23@2x" src="https://user-images.githubusercontent.com/479169/229543303-1a114b45-4313-4f55-8916-ce8270622185.png">

bendyorke replied:
This is a scrollbar, some work before launch needs to be done to remove it (it's visually unpleasant but for accessibility reasons we shouldn't just entirely disable it). I'd like to move that to an improvement before launching v2 as the default (should not block merge)

tiensonqin commented on deps/shui/shui-graph/logseq/bak/pages/shui___components___table/2023-03-05T11_28_36.962Z.Desktop.md:
```diff
@@ -0,0 +1,55 @@
+- ### Props
```
Should those backup files be removed?

tiensonqin commented on deps/shui/shui-graph/pages/About Shui.md:
```diff
@@ -0,0 +1,22 @@
+- ## What is shui?
+- Shui is the component library for logseq. It has 3 main goals:
```
🥳

tiensonqin reviewed (approved):
So exciting for `shui`, a massive step for the core UI components! 🎉 🚢 
I left some minor feedback.

logseq-cldwalker commented on deps/shui/shui-graph/pages/shui___components___table.md:
```diff
@@ -0,0 +1,61 @@
+		  | 123 | 从学生到研究人员，Logseq 灵活直观的界面使其成为任何想要优化笔记和知识管理工作流程的人的理想工具。|  | x |
+		  | 152 | 无论您是想整理自己的思路、与他人合作，还是简化笔记流程，Logseq 提供的革命性方法肯定会改变您的工作方式。|  | x |
+	- #### Query table for blocks
+		- {{query #table-example/block}}
```
When using the query builder, I noticed that the property filter now displays the new built-in properties. My guess is users are not going to want this but we can also wait and see:
<img width="225" alt="Screen Shot 2023-04-03 at 5 57 41 PM" src="https://user-images.githubusercontent.com/97210743/229636887-fad93ceb-2995-4c35-93d0-9920d1493748.png">

This happens because these properties haven't been added to `gp-property/hidden-built-in-properties`

bendyorke replied:
I added them to `editable-built-in-properties`, as they should not be hidden. They still appear in the query-builder. 

If you'd like, immediately after merge I can work on a PR to hide them from the query builder, but it should not block this PR

logseq-cldwalker reviewed (approved):
@bendyorke Thanks for kicking off a beautiful, reusable component and library! 👍 🕺 I tried a couple of the options and they all worked as described. Left a couple minor things to fix. If `gp-property/hidden-built-in-properties` ends up being a rabbit hole, I can also do it as a followup after this is merged

tiensonqin commented:
@bendyorke anything we can help to get this PR merged?

bendyorke commented on src/main/frontend/components/query_table.cljs:
```diff
@@ -165,52 +167,66 @@
+          table-version (get-shui-component-version :table config)
+          result-as-text (for [result-entry result'] 
+                           (for [column-name columns] 
+                             (case column-name 
```
The reason being is that I need access to a text representation of how the column will be rendered (number of characters). Support for more complex table column types (images, assets, etc) can come in the future.

I tried to get it to work using the existing `build-column-value` but you lose a lot. I'll make sure to mention this in any release comms--I do expect some tables to look different, and I'm happy to tackle those as they come in and make sure they're properly supported.

bendyorke replied:
I've refactored it for now to better highlight the difference in functionality

bendyorke commented on deps/shui/shui-graph/pages/shui___components___table.md:
```diff
@@ -0,0 +1,61 @@
+	  logseq.table.stripes:: true
+	  | Prop Name | Description | Values |
+	  | --- | --- | --- |
+	  | `logseq.table.version` | The version of the table | 1, 2 |
```
I added them to editable-properties, they should not be hidden properties. Likely we will have to introduce another tier of properties, we can do that before launching this as the default version is that's a blocker for you. Shouldn't block this initial PR though

logseq-cldwalker replied:
Looks like these were added to `hidden-built-in-properties`. I'm guessing that's a mistake as they wouldn't be editable

bendyorke replied:
Oh wow good catch. My git's being weird, they were under `editable-` when I opened the file, but I think I've fixed it now. I'll check before deploy, thanks again
